src/OpenApi/src/Extensions/JsonObjectSchemaExtensions.cs

```diff
@@ -0,0 +1,279 @@
+            {
+                schema["type"] = "integer";
+                schema["format"] = constraint is LongRouteConstraint ? "int64" : "int32";
+            }
+            else if (constraint is GuidRouteConstraint or StringRouteConstraint)
+            {
+                schema["type"] = "string";
+                schema["format"] = constraint is GuidRouteConstraint ? "uuid" : null;
+            }
+            else if (constraint is BoolRouteConstraint)
+            {
+                schema["type"] = "boolean";
+                schema["format"] = null;
+            }
+            else if (constraint is AlphaRouteConstraint)
+            {
+                schema["type"] = "string";
+                schema["format"] = null;
+            }
+            else if (constraint is DateTimeRouteConstraint)
+            {
+                schema["type"] = "string";
+                schema["format"] = "date-time";
+            }
+        }
+    }
+
+    /// <summary>
+    /// Applies parameter-specific customizations to the target schema.
+    /// </summary>
+    /// <param name="schema">The <see cref="JsonObject"/> produced by the underlying schema generator.</param>
+    /// <param name="parameterDescription">The <see cref="ApiParameterDescription"/> associated with the <see paramref="schema"/>.</param>
+    internal static void ApplyParameterInfo(this JsonObject schema, ApiParameterDescription parameterDescription)
+    {
+        // Route constraints are only defined on parameters that are sourced from the path. Since
+        // they are encoded in the route template, and not in the type information based to the underlying
+        // schema generator we have to handle them separately here.
+        if (parameterDescription.RouteInfo?.Constraints is { } constraints)
+        {
+            schema.ApplyRouteConstraints(constraints);
+        }
+        // This is special handling for parameters that are not bound from the body but represented in a complex type.
+        // For example:
+        //
+        // public class MyArgs
+        // {
+        //     [Required]
+        //     [Range(1, 10)]
+        //     [FromQuery]
+        //     public string Name { get; set; }
+        // }
+        //
+        // public IActionResult(MyArgs myArgs) { }
+        //
+        // In this case, the `ApiParameterDescription` object that we received will represent the `Name` property
+        // based on our model binding heuristics. In that case, to access the validation attributes that the
+        // model binder will respect we will need to get the property from the container type and map the
+        // attributes on it to the schema.
+        if (parameterDescription.ModelMetadata.PropertyName is { } propertyName)
+        {
+            var property = parameterDescription.ModelMetadata.ContainerType?.GetProperty(propertyName, BindingFlags.Public | BindingFlags.Instance);
```

Comment:
Are private/internal properties deserialized by Json? Also, Json can use the constructor for deserializing, those parameters can have attributes as well right?

Reply:
> Are private/internal properties deserialized by Json? 

Nope.

> Also, Json can use the constructor for deserializing, those parameters can have attributes as well right?

The gotcha here is that this isn't actually JSON. We're trying to handle the scenario where parameters that are serialized from the route/querystring are bound as properties on a model. They don't go through the MVC model binding layer and not the JSON serialization layer. That's why we apply this change _after_ schema generation has run.